source/vibe/core/sync.d

```diff
@@ -819,3 +819,490 @@ private struct TaskConditionImpl(bool INTERRUPTIBLE, LOCKABLE) {
+ * 
+ *  The class is implemented as a template with an empty parameter list 
+ *  to allow the automatical inference of shared members. To instantiate it,
+ *  use the alias $(D TaskReadWriteMutex) defined in this module.
+ * 
+ *  Notice:
+ *      Mutexes implemented by this class cannot be interrupted
+ *      using $(D vibe.core.task.Task.interrupt()). The corresponding
+ *      InterruptException will be deferred until the next blocking
+ *      operation yields the event loop.
+ * 
+ *      Use $(D InterruptibleTaskReadWriteMutex) as an alternative that can be
+ *      interrupted.
+ * 
+ *  cf. $(D core.sync.mutex.ReadWriteMutex)
+ */
+class TaskReadWriteMutexImpl()
+{
+    private {
+        alias State = ReadWriteMutexState!false;
+        alias LockingIntent = State.LockingIntent;
+        alias READ_ONLY  = LockingIntent.READ_ONLY;
+        alias READ_WRITE = LockingIntent.READ_WRITE;
+        
+        /** The shared state used by the reader and writer mutexes. */
+        State m_state;
+    }
+    
+    /** The policy with which the mutex should operate.
+     *
+     *  The policy determines how the acquisition of the locks is 
+     *  performed and can be used to tune the mutex according to the
+     *  underlying algorithm in which it is used.
+     *
+     *  According to the provided policy, the mutex will either favor
+     *  reading or writing tasks and could potentially starve the 
+     *  respective opposite.
+     *
+     *  cf. $(D core.sync.rwmutex.ReadWriteMutex.Policy)
+     */
+    alias Policy = State.Policy;
+    
+    /** A common baseclass for both of the provided mutexes.
+     *
+     *  The intent for the according mutex is specified through the 
+     *  $(D INTENT) template argument, which determines if a mutex is 
+     *  used for read or write locking.
+     */
+    final class Mutex(LockingIntent INTENT): core.sync.mutex.Mutex, Lockable
+    {
+        /** Try to lock the mutex. cf. $(D core.sync.mutex.Mutex) */
+        override bool tryLock() { return m_state.tryLock!INTENT(); }
+        /** Lock the mutex. cf. $(D core.sync.mutex.Mutex) */
+        override void lock()    { m_state.lock!INTENT(); }
+        /** Unlock the mutex. cf. $(D core.sync.mutex.Mutex) */
+        override void unlock()  { m_state.unlock!INTENT(); }
+    }
+    alias Reader = Mutex!READ_ONLY;
+    alias Writer = Mutex!READ_WRITE;
+    
+    Reader reader;
+    Writer writer;
+    
+    this(Policy policy = Policy.PREFER_WRITERS)
+    {
+        m_state = State(policy);
+        reader = new Reader();
+        writer = new Writer();
+    }
+    
+    /** The policy with which the lock has been created. */
+    @property Policy policy() const { return m_state.policy; }
+}
+
+/** A simple alias for $(D TaskReadWriteMutexImpl).
+ *
+ *  Use this alias to avoid providing an empty template parameter list when
+ *  instantiating a mutex.
+ */
+alias TaskReadWriteMutex = TaskReadWriteMutexImpl!();
```

Comment:
As noted, adding empty template parameter lists does indeed allow the compiler to infer shared methods again. The downside is, that for this to work, each mutex would have to be initialized using an explicit empty parameter list, e.g. `TaskReadWriteMutex!()`.

To mitigate this, i decided to name the classes differently (`TaskReadWriteMutexImpl!()`) but leave them public to guarantee useful documentation. The mutex type itself can be referred to through an alias (i.e. `TaskReadWriteMutex x();` still works).